crates/core/tedge_agent/src/operation_file_cache/mod.rs

```diff
@@ -0,0 +1,349 @@
+use tedge_actors::Builder;
+use tedge_actors::DynSender;
+use tedge_actors::LinkError;
+use tedge_actors::LoggingReceiver;
+use tedge_actors::MessageReceiver;
+use tedge_actors::MessageSink;
+use tedge_actors::NoConfig;
+use tedge_actors::RuntimeError;
+use tedge_actors::RuntimeRequest;
+use tedge_actors::RuntimeRequestSink;
+use tedge_actors::Sender;
+use tedge_actors::ServiceProvider;
+use tedge_actors::SimpleMessageBoxBuilder;
+use tedge_api::messages::ConfigUpdateCmdPayload;
+use tedge_api::mqtt_topics::Channel;
+use tedge_api::mqtt_topics::ChannelFilter;
+use tedge_api::mqtt_topics::EntityFilter;
+use tedge_api::mqtt_topics::EntityTopicId;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::OperationType;
+use tedge_api::path::DataDir;
+use tedge_api::CommandStatus;
+use tedge_downloader_ext::DownloadRequest;
+use tedge_downloader_ext::DownloadResult;
+use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
+use tedge_mqtt_ext::TopicFilter;
+use tracing::error;
+use tracing::info;
+use tracing::warn;
+
+type IdDownloadRequest = (String, DownloadRequest);
+type IdDownloadResult = (String, DownloadResult);
+
+fan_in_message_type!(FileCacheInput[MqttMessage, IdDownloadResult]: Debug);
+
+pub struct FileCacheActor {
+    input_receiver: LoggingReceiver<FileCacheInput>,
+    mqtt_sender: DynSender<MqttMessage>,
+    downloader_sender: DynSender<IdDownloadRequest>,
+
+    tedge_http_host: Arc<str>,
+    mqtt_schema: MqttSchema,
+
+    pending_operations: HashMap<String, ConfigUpdateCmdPayload>,
+}
+
+#[async_trait]
+impl Actor for FileCacheActor {
+    fn name(&self) -> &str {
+        "FileCacheActor"
+    }
+
+    async fn run(mut self) -> Result<(), RuntimeError> {
+        while let Some(message) = self.input_receiver.recv().await {
+            match message {
+                FileCacheInput::MqttMessage(message) => {
+                    self.process_mqtt_message(message).await?;
+                }
+                FileCacheInput::IdDownloadResult(message) => {
+                    self.process_download(message).await?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+impl FileCacheActor {
+    async fn process_mqtt_message(
+        &mut self,
+        mqtt_message: MqttMessage,
+    ) -> Result<(), RuntimeError> {
+        let Ok((entity, channel)) = self.mqtt_schema.entity_channel_of(&mqtt_message.topic) else {
+            return Ok(());
+        };
+
+        let Channel::Command {
+            operation: OperationType::ConfigUpdate,
+            cmd_id,
+        } = channel
+        else {
+            return Ok(());
+        };
+
+        let Ok(update_payload) =
+            serde_json::from_slice::<ConfigUpdateCmdPayload>(mqtt_message.payload.as_bytes())
+        else {
+            warn!("Received config update, but payload is malformed");
+            return Ok(());
+        };
+
+        if update_payload.remote_url.is_empty() || update_payload.tedge_url.is_some() {
```

Comment:
I'm hypothesising a lot here, but it feels here like there is some logic that decides from the payload what state we're in that's distributed around this actor's methods. Can we do some sort of conversion from `ConfigUpdateCmdPayload` into an enum that represents what state we're currently in so we process that information in a single place?

Reply:
I may be missing what you mean slightly, but IMO this is much more the case in other operation handlers, which have more states, here we only: 1. process messages which have a `remoteUrl` and don't have `tedgeUrl`, 2. download from the remote URL, put in the file-transfer directory, and put the URL inside `tedgeUrl`. I agree that transitions between states in operation handling is hardly visible, but I thought about improving that in all the operations later, but here I'm not sure if you mean something simpler than that.

Reply:
I hadn't deeply considered a solution, it just "felt like" it could do with being more "parse, don't validate", but I'm happy to be overruled here.

Reply:
My main concern is that the logic to process a config update request is scattered in so many places that it's really difficult to be sure that this is working as expected and really easy to break.

- Instead of what we have today (some dumb payloads in `tedge_api` and processing logic in `c8y_mapper_ext`, `tedge_config_ext` and now `tedge_agent`)
- I would prefer a central place where are defined the misc steps required to process a configuration update and to reduce the role of the mapper, agent, and plugins to trigger actions.

This is how I understand @jarhodes314 call for "parse, don't validate":

```
impl FileCacheActor {
    async fn process_mqtt_message(
        &mut self,
        mqtt_message: MqttMessage,
    ) -> Result<(), RuntimeError> {
        match ConfigUpdate::action_for(mqtt_message) {
            Some(DownloadRemote { cloud_url, }) => {
                ...
            } 
        }
}
```

Unfortunately, this is not the best time to fix that.